File: tokens/token-swap/steel/tests/instruction/create_amm.ts
import { PublicKey, SystemProgram, TransactionInstruction } from '@solana/web3.js';
import * as borsh from 'borsh';
import { Assignable, TokenSwapInstruction } from './instruction';

class CreateAmm extends Assignable {
  toBuffer() {
    return Buffer.from(borsh.serialize(CreateAmmSchema, this));
  }
}

const CreateAmmSchema = new Map([
  [
    CreateAmm,
    {
      kind: 'struct',
      fields: [
        ['instruction', 'u8'],
        ['id', [32]],
        ['fee', 'u16'],
      ],
    },
  ],
]);

export function buildCreateAmmInstruction(props: {
  id: PublicKey;
  fee: number;
  amm: PublicKey;
  admin: PublicKey;
  payer: PublicKey;
  programId: PublicKey;
}) {
  const ix = new CreateAmm({
    instruction: TokenSwapInstruction.CreateAmm,
    id: props.id.toBytes(),
    fee: props.fee,
  });

  return new TransactionInstruction({
    keys: [
      {
        pubkey: props.amm,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: props.admin,
        isSigner: false,
        isWritable: true,
      },
      {
        pubkey: props.payer,
        isSigner: true,
        isWritable: true,
      },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
    programId: props.programId,
    data: ix.toBuffer(),
  });
}

File: tokens/token-swap/steel/program/src/instructions/create_amm.rs
use crate::{state::*, SteelInstruction};
use steel::*;

instruction!(SteelInstruction, CreateAmm);

#[repr(C, packed)]
#[derive(Clone, Copy, Debug, Pod, Zeroable)]
pub struct CreateAmm {
    id: Pubkey,
    fee: u16,
}

impl CreateAmm {
    pub fn process(
        program_id: &Pubkey,
        accounts: &[AccountInfo<'_>],
        data: &[u8],
    ) -> ProgramResult {
        let args = Self::try_from_bytes(data)?;

        let [amm_info, admin, payer, system_program] = accounts else {
            return Err(ProgramError::NotEnoughAccountKeys);
        };

        payer.is_writable()?;
        system_program.is_program(&system_program::ID)?;

        // create amm
        create_account::<Amm>(
            amm_info,
            system_program,
            payer,
            program_id,
            &[args.id.as_ref()],
        )?;

        let amm = amm_info.as_account_mut::<Amm>(program_id)?;

        amm.admin = *admin.key;
        amm.fee = args.fee;
        amm.id = args.id;

        Ok(())
    }
}
